neqo-transport/src/connection.rs

```diff
@@ -634,88 +633,106 @@ impl Connection {
-                    // but a non-zero maximum tells us if we are accepting 0-RTT.
-                    self.zero_rtt_state = if self.crypto.tls.preinfo()?.max_early_data() > 0 {
-                        ZeroRttState::Accepted
-                    } else {
-                        ZeroRttState::Rejected
-                    };
-                } else {
-                    qdebug!([self] "Changing to use Server CID={}", hdr.scid.as_ref().unwrap());
-                    let p = self
-                        .paths
-                        .iter_mut()
-                        .find(|p| p.received_on(&d))
-                        .expect("should have a path for sending Initial");
-                    p.remote_cid = hdr.scid.unwrap();
+    fn decrypt_body(&mut self, mut hdr: &mut PacketHdr, slc: &[u8]) -> Option<Vec<u8>> {
+        // Decryption failure, or not having keys is not fatal.
+        // If the state isn't available, or we can't decrypt the packet, drop
+        // the rest of the datagram on the floor, but don't generate an error.
+        let largest_acknowledged = self
+            .loss_recovery
+            .largest_acknowledged(PNSpace::from(hdr.epoch));
+        let res = match self.crypto.obtain_crypto_state(self.role, hdr.epoch) {
+            Ok(cs) => match cs.rx.as_ref() {
+                Some(rx) => {
+                    let pn_decoder = PacketNumberDecoder::new(largest_acknowledged);
+                    decrypt_packet(rx, pn_decoder, &mut hdr, slc)
                 }
-                self.set_state(State::Handshaking);
-            }
+                _ => Err(Error::KeysNotFound),
+            },
+            Err(e) => Err(e),
+        };
+        res.ok()
+    }
 
-            if !self.paths.iter().any(|p| p.received_on(&d)) {
-                // Right now, we don't support any form of migration.
-                // So generate an error if a packet is received on a new path.
-                return Err(Error::InvalidMigration);
-            }
+    /// Ok(true) if the packet is a duplicate
+    fn process_ack(&mut self, hdr: &PacketHdr, body: Vec<u8>, now: Instant) -> Res<bool> {
+        // TODO(ekr@rtfm.com): Have the server blow away the initial
+        // crypto state if this fails? Otherwise, we will get a panic
+        // on the assert for doesn't exist.
+        // OK, we have a valid packet.
+
+        // TODO(ekr@rtfm.com): Filter for valid for this epoch.
+
+        let ack_eliciting = self.input_packet(hdr.epoch, Decoder::from(&body[..]), now)?;
+        let space = PNSpace::from(hdr.epoch);
+        if self.acks[space].is_duplicate(hdr.pn) {
+            qdebug!([self] "Received duplicate packet epoch={} pn={}", hdr.epoch, hdr.pn);
+            self.stats.dups_rx += 1;
+            Ok(true)
+        } else {
+            self.acks[space].set_received(now, hdr.pn, ack_eliciting);
+            Ok(false)
         }
+    }
 
+    fn process_handshake(&mut self, hdr: PacketHdr, d: &Datagram) -> Res<()> {
```

Comment:
The purpose of this function is to update connections IDs.  It also sets zero_rtt_state, which seems a little different.  Maybe name it `start_handshake()` or something like that.